### ISSUE TITLE
Introduce an `%import` pragma

### ISSUE THREAD

**Issue by jubianchi** [hoaproject/Compiler#54]
While working on [TML](https://github.com/jubianchi/tml) I needed to write a second grammar for test purpose. This grammar is quite the same as the original language grammar but with some rules overloaded.

I played a bit whit compiler and came up with an `%import` directive. It works quite well in my case and was wondering if you would want such feature in compiler. Here is how it looks like:

```
// src/tml.pp

%token             T_FN                    \.[a-zA-Z_][a-zA-Z0-9_]*
%token             T_VAR                   @[a-zA-Z_][a-zA-Z0-9_]*
%token             T_NUMBER                \-?[1-9][0-9]*

#tml:
    ( fn() | expr() | assign() | str() )+

#expr:
    ( <T_NUMBER> | rvar() ) ( operator() expr() )?

// ...
```

```
// tests/tml.pp

%import ../src/tml.pp

#tml:
      expr()
    | assign()
    | str()

#expr:
    <T_NUMBER> ( operator() expr() | division() )?

// Avoids division by zero
division:
    <T_NUMBER[0]> <T_OP_DIVIDE> <T_NUMBER[0]>

operator:
      <T_OP_PLUS>
    | <T_OP_MINUS>
    | <T_OP_MULTI>

```

Basically, the `tests/tml.pp` will be loaded (`Compiler\Llk\Llk::load`) and parsed. The `%import` directive will be reached and the `src/tml.pp` grammar will in turn be loaded and parsed. Then we continue with `tests/tml.pp` thus overloading rules and tokens from `src/tml.pp`.

The `%import` can be used at the top of the file to overload imported grammar or a anywhere else to produce different results. Grammar are imported relatively to the file importing it. If an imported grammar imports in turn another grammar, the import will be relative to the file where the `%import` directive is written.

What do you think ?

**Comment by Hywan** [comment]
Hello :-),

During my PhD thesis, I mentionned this feature and addressing it is a very challenging task. You don't only **import** rules here, you are **extending** them or even worse: You are also **overriding** them. This is difficult. Hardly difficult. Extending would be difficult, refining would be much more difficult, but overriding… damn. This is too error-prone because: What if you are depending on a grammar that radically changes? Your grammar will be broken. We have to detect this, and this is **very** difficult.

**Comment by Hywan** [comment]
So, it would be :-1: for me regarding the amount of work. But, let's the community decides. If this is something the community asks, I could work on it. Don't expect this feature to be availble in a week. Just the formalization could take months.

**Comment by jubianchi** [comment]
@Hywan fine ;) I already reverted my code in tml. 

TBH I'm not aware of the problems but if you can explain them or give me a link to something giving more clue, I would be glad to work on this.

**Comment by CircleCode** [comment]
I really love the idea, but I understand how hard this could be.
Maybe some mitigation could be some kind of
`%import: path/to/file.pp:%token|#rule`
that would just work like computer-assisted copy paste (are you thinking about _traits_?)

This way, no overriding, no extending, and no huge copy paste, only a big list of imports, and some custom rules.

What do you think about this?

Note: I understand this would lead to pp files with a lot of imports and only some custom rules, making files much much bigger than a full gramma import, but the idea here is not to reduce grammar length, only to avoid huge parts to be duplicated.

On the parsing side, I think (but I do not know the compiler internals) that this does not change radically the way compiler works

**Comment by jubianchi** [comment]
@CircleCode I was also thinking of `%import`ing only tokens or only rules but would propose it in a second time, but I was probably wrong.

A first step could be to support on token imports but we still have to take care of some things: if the imported grammar changes one of its token regex, the importing grammar could break.

**Comment by CircleCode** [comment]
@jubianchi right, but the break would be easy to detect since once the replacement (aka _) has been done, the grammar should be good on its own

**Comment by stephpy** [comment]
It's really interesting, but I understand @Hywan, it seems hard ...

- We could assume than `%import%` does like an include in PHP, if you override tokens, rules, ... the 2nd occurence override the 1st.
- We could use `%import%` to load in background, then `use xx_token; use yy_token; etc...`, It would be more flexible but harder to implement for client.

**Comment by 1e1** [comment]
Sorry for my dummy question. 
Before building the AST, a file is like a key-value of "lines"? 
Can we rebuild a cached .pp which is a merge of both files. 

(I'm always curious about the context leading to write a new grammar).

**Comment by Hywan** [comment]
@1e1 I don't understand your comment. A grammar is a set of lexemes and a set of rules. This is not simply “merging” them. A grammar describes a syntax. Merging syntax and still be coherent or detect conflict is not simple. Actually, I don't know if this is even automatically done.
